File: backend/app/models/opportunity.py
"""Opportunity model — represents a customer opportunity (商机线索)"""
from typing import Optional
from sqlalchemy import String, Text, JSON
from sqlalchemy.orm import Mapped, mapped_column
from .base import Base


class Opportunity(Base):
    __tablename__ = "opportunities"
    __table_args__ = {"schema": "opportunities"}

    opportunity_id: Mapped[str] = mapped_column(String, primary_key=True)
    customer_name: Mapped[Optional[str]] = mapped_column(String, default=None)
    sales_person: Mapped[Optional[str]] = mapped_column(String, default=None)
    fae: Mapped[Optional[str]] = mapped_column(String, default=None)
    quotation_person: Mapped[Optional[str]] = mapped_column(String, default=None)
    # 从 Quotation 迁移的商机级字段
    platform_type: Mapped[Optional[str]] = mapped_column(String, default=None)
    chassis_form: Mapped[Optional[str]] = mapped_column(String, default=None)
    # D1 商机结果与复盘（蓝图 A1-A2）— 解锁 M4 丢标复盘 / P2 直销渠道双基线 / M1 行业打法
    industry: Mapped[Optional[str]] = mapped_column(String, default=None)        # 行业（教育/政府/金融/制造…）
    customer_type: Mapped[Optional[str]] = mapped_column(String, default=None)   # 客户类型（直销/渠道/集成商/最终用户）
    result: Mapped[str] = mapped_column(String, default="pending")               # 业务结果：pending / won / lost（与 status 正交）
    purchase_qty: Mapped[int] = mapped_column(default=0)
    created_at: Mapped[Optional[str]] = mapped_column(String, default=None)
    updated_at: Mapped[Optional[str]] = mapped_column(String, default=None)
    status: Mapped[Optional[str]] = mapped_column(String, default="active")
    extra_fields: Mapped[Optional[str]] = mapped_column(Text, default=None)  # JSON
    tenant_id: Mapped[Optional[str]] = mapped_column(String, default="default")

    def to_dict(self) -> dict:
        import json
        result = {
            "opportunity_id": self.opportunity_id,
            "customer_name": self.customer_name or "",
            "sales_person": self.sales_person or "",
            "fae": self.fae or "",
            "quotation_person": self.quotation_person or "",
            "platform_type": self.platform_type or "",
            "chassis_form": self.chassis_form or "",
            "industry": self.industry or "",
            "customer_type": self.customer_type or "",
            "result": self.result or "pending",
            "purchase_qty": self.purchase_qty or 0,
            "created_at": self.created_at or "",
            "updated_at": self.updated_at or "",
            "status": self.status or "active",
            "tenant_id": self.tenant_id or "default",
        }
        # 展开 extra_fields JSON 到顶层
        if self.extra_fields:
            try:
                extra = json.loads(self.extra_fields)
                result.update(extra)
            except (json.JSONDecodeError, TypeError):
                pass
        return result

File: frontend/src/types/opportunity.ts
/**
 * Project and Quotation type definitions
 */

export interface Project {
  opportunity_id: string
  customer_name: string
  status: string
  created_at: string
  updated_at: string
  
  // Project-level metadata (user-editable)
  purchase_qty: number
  platform_type: string
  chassis_form: string
  sales_person: string
  fae: string
  // 商机上下文（pricing scope 维度，可选 — 后端 extra_fields 动态字段）
  customer_type?: string
  // 业务结果流转（与 status 正交）
  result?: 'pending' | 'won' | 'lost'

  // Computed fields from latest quotation
  quotation_count: number
  config_count: number
}

export interface Quotation {
  quotation_id: string
  opportunity_id: string
  version: string
  quotation_name: string
  file_path?: string

  // Quotation-level fields (user-editable)
  quotation_date: string

  // Computed totals
  l6_price: number
  total_qty: number
  config_count: number
  total_price: number
  profit_margin: number

  created_at: string
  updated_at: string
  status: string
  // 草稿/已导出状态机：NULL=草稿(可进工作台)；非空=已导出冻结(时间戳，点列表只看 Excel+成本)
  exported_at?: string | null
  cost_snapshot?: Record<string, any> | null
  // 列表轻量标志（list 端点剥离完整 snapshot，留布尔供行内判断）
  has_cost_snapshot?: boolean
  // 手工补录过（manual:true，未冻结）→ 列表给「编辑成本」入口可二次进抽屉改
  has_manual_cost?: boolean

  // WIP fields (primary-quotation flag + platform classification)
  is_primary?: boolean
  platform_type?: string
}

// Alias: Project is being renamed to Opportunity (商机) across the UI.
// Kept as a type alias so views can use the semantically-correct name
// without duplicating the shape.
export type Opportunity = Project

export interface ProjectItem {
  item_id?: number
  quotation_id: string
  config_name: string
  category: string
  catalogue: string
  description: string
  part_category: string
  qty: number
  base_price: number
  final_price: number
  profit_margin: number
}

export interface ProjectListResponse {
  items: Project[]
  total: number
}

export interface QuotationListResponse {
  quotations: Quotation[]
}
